diff --git a/common_riscv64.h b/common_riscv64.h
index 7ddbe80..3998cd2 100644
--- a/common_riscv64.h
+++ b/common_riscv64.h
@@ -91,8 +91,109 @@ static inline int blas_quickdivide(blasint x, blasint y){
 #define BUFFER_SIZE     ( 32 << 20)
 #define SEEK_ADDRESS

-#if defined(C910V)
+#if defined(C910V) || defined(__riscv_vector)
 #include <riscv_vector.h>
 #endif

+#if defined(__riscv_vector)
+#define vsetvl_e32m4 __riscv_vsetvl_e32m4
+#define vsetvl_e32m8 __riscv_vsetvl_e32m8
+#define vsetvl_e64m4 __riscv_vsetvl_e64m4
+#define vsetvl_e64m8 __riscv_vsetvl_e64m8
+#define vsetvlmax_e32m1 __riscv_vsetvlmax_e32m1
+#define vsetvlmax_e64m1 __riscv_vsetvlmax_e64m1
+
+#define vle32_v_f32m4 __riscv_vle32_v_f32m4
+#define vle32_v_f32m8 __riscv_vle32_v_f32m8
+#define vle64_v_f64m4 __riscv_vle64_v_f64m4
+#define vle64_v_f64m8 __riscv_vle64_v_f64m8
+#define vlse32_v_f32m4 __riscv_vlse32_v_f32m4
+#define vlse32_v_f32m8 __riscv_vlse32_v_f32m8
+#define vlse64_v_f64m4 __riscv_vlse64_v_f64m4
+#define vlse64_v_f64m8 __riscv_vlse64_v_f64m8
+#define vse32_v_f32m4 __riscv_vse32_v_f32m4
+#define vse32_v_f32m8 __riscv_vse32_v_f32m8
+#define vse64_v_f64m4 __riscv_vse64_v_f64m4
+#define vse64_v_f64m8 __riscv_vse64_v_f64m8
+#define vsse32_v_f32m4 __riscv_vsse32_v_f32m4
+#define vsse32_v_f32m8 __riscv_vsse32_v_f32m8
+#define vsse64_v_f64m4 __riscv_vsse64_v_f64m4
+#define vsse64_v_f64m8 __riscv_vsse64_v_f64m8
+
+#define vfmacc_vf_f32m4 __riscv_vfmacc_vf_f32m4
+#define vfmacc_vf_f64m4 __riscv_vfmacc_vf_f64m4
+#define vfmacc_vv_f32m4 __riscv_vfmacc_vv_f32m4
+#define vfmacc_vv_f64m4 __riscv_vfmacc_vv_f64m4
+#define vfmul_vf_f32m4 __riscv_vfmul_vf_f32m4
+#define vfmul_vf_f32m8 __riscv_vfmul_vf_f32m8
+#define vfmul_vf_f64m4 __riscv_vfmul_vf_f64m4
+#define vfmul_vf_f64m8 __riscv_vfmul_vf_f64m8
+#define vfmul_vv_f32m4 __riscv_vfmul_vv_f32m4
+#define vfmul_vv_f64m4 __riscv_vfmul_vv_f64m4
+#define vfadd_vv_f32m8 __riscv_vfadd_vv_f32m8
+#define vfadd_vv_f64m8 __riscv_vfadd_vv_f64m8
+#define vfdiv_vf_f32m4 __riscv_vfdiv_vf_f32m4
+#define vfdiv_vf_f64m4 __riscv_vfdiv_vf_f64m4
+#define vfmax_vv_f32m8 __riscv_vfmax_vv_f32m8
+#define vfmax_vv_f64m8 __riscv_vfmax_vv_f64m8
+#define vfmin_vv_f32m8 __riscv_vfmin_vv_f32m8
+#define vfmin_vv_f64m8 __riscv_vfmin_vv_f64m8
+
+#define vfmv_v_f_f32m1 __riscv_vfmv_v_f_f32m1
+#define vfmv_v_f_f32m4 __riscv_vfmv_v_f_f32m4
+#define vfmv_v_f_f32m8 __riscv_vfmv_v_f_f32m8
+#define vfmv_v_f_f64m1 __riscv_vfmv_v_f_f64m1
+#define vfmv_v_f_f64m4 __riscv_vfmv_v_f_f64m4
+#define vfmv_v_f_f64m8 __riscv_vfmv_v_f_f64m8
+#define vfmv_f_s_f32m1_f32 __riscv_vfmv_f_s_f32m1_f32
+#define vfmv_f_s_f32m4_f32 __riscv_vfmv_f_s_f32m4_f32
+#define vfmv_f_s_f64m1_f64 __riscv_vfmv_f_s_f64m1_f64
+#define vfmv_f_s_f64m4_f64 __riscv_vfmv_f_s_f64m4_f64
+
+#define vfredmax_vs_f32m4_f32m1(acc, vec, scalar, vl) __riscv_vfredmax_vs_f32m4_f32m1(vec, scalar, vl)
+#define vfredmax_vs_f32m8_f32m1(acc, vec, scalar, vl) __riscv_vfredmax_vs_f32m8_f32m1(vec, scalar, vl)
+#define vfredmax_vs_f64m4_f64m1(acc, vec, scalar, vl) __riscv_vfredmax_vs_f64m4_f64m1(vec, scalar, vl)
+#define vfredmax_vs_f64m8_f64m1(acc, vec, scalar, vl) __riscv_vfredmax_vs_f64m8_f64m1(vec, scalar, vl)
+#define vfredmin_vs_f32m8_f32m1(acc, vec, scalar, vl) __riscv_vfredmin_vs_f32m8_f32m1(vec, scalar, vl)
+#define vfredmin_vs_f64m8_f64m1(acc, vec, scalar, vl) __riscv_vfredmin_vs_f64m8_f64m1(vec, scalar, vl)
+#define vfredosum_vs_f32m4_f32m1(acc, vec, scalar, vl) __riscv_vfredosum_vs_f32m4_f32m1(vec, scalar, vl)
+#define vfredosum_vs_f32m8_f32m1(acc, vec, scalar, vl) __riscv_vfredosum_vs_f32m8_f32m1(vec, scalar, vl)
+#define vfredosum_vs_f64m4_f64m1(acc, vec, scalar, vl) __riscv_vfredosum_vs_f64m4_f64m1(vec, scalar, vl)
+#define vfredusum_vs_f32m4_f32m1(acc, vec, scalar, vl) __riscv_vfredusum_vs_f32m4_f32m1(vec, scalar, vl)
+#define vfredusum_vs_f64m4_f64m1(acc, vec, scalar, vl) __riscv_vfredusum_vs_f64m4_f64m1(vec, scalar, vl)
+#define vfredusum_vs_f64m8_f64m1(acc, vec, scalar, vl) __riscv_vfredusum_vs_f64m8_f64m1(vec, scalar, vl)
+
+#define vmflt_vf_f32m4_b8 __riscv_vmflt_vf_f32m4_b8
+#define vmflt_vf_f32m8_b4 __riscv_vmflt_vf_f32m8_b4
+#define vmflt_vf_f64m4_b16 __riscv_vmflt_vf_f64m4_b16
+#define vmflt_vf_f64m8_b8 __riscv_vmflt_vf_f64m8_b8
+#define vmflt_vv_f32m8_b4 __riscv_vmflt_vv_f32m8_b4
+#define vmflt_vv_f64m8_b8 __riscv_vmflt_vv_f64m8_b8
+#define vmfgt_vf_f32m4_b8 __riscv_vmfgt_vf_f32m4_b8
+#define vmfgt_vf_f64m4_b16 __riscv_vmfgt_vf_f64m4_b16
+#define vmfge_vf_f32m8_b4 __riscv_vmfge_vf_f32m8_b4
+#define vmfge_vf_f64m8_b8 __riscv_vmfge_vf_f64m8_b8
+#define vmfle_vf_f32m8_b4 __riscv_vmfle_vf_f32m8_b4
+#define vmfle_vf_f64m8_b8 __riscv_vmfle_vf_f64m8_b8
+#define vmfirst_m_b4 __riscv_vfirst_m_b4
+#define vmfirst_m_b8 __riscv_vfirst_m_b8
+#define vmfirst_m_b16 __riscv_vfirst_m_b16
+
+#define vfrsub_vf_f32m4_m(mask, maskedoff, op1, op2, vl) __riscv_vfrsub_vf_f32m4_m(mask, op1, op2, vl)
+#define vfrsub_vf_f32m8_m(mask, maskedoff, op1, op2, vl) __riscv_vfrsub_vf_f32m8_m(mask, op1, op2, vl)
+#define vfrsub_vf_f64m4_m(mask, maskedoff, op1, op2, vl) __riscv_vfrsub_vf_f64m4_m(mask, op1, op2, vl)
+#define vfrsub_vf_f64m8_m(mask, maskedoff, op1, op2, vl) __riscv_vfrsub_vf_f64m8_m(mask, op1, op2, vl)
+
+#define vmv_v_x_u32m8 __riscv_vmv_v_x_u32m8
+#define vmv_v_x_u64m8 __riscv_vmv_v_x_u64m8
+#define vid_v_u32m8 __riscv_vid_v_u32m8
+#define vid_v_u64m8 __riscv_vid_v_u64m8
+#define vid_v_u32m8_m(mask, maskedoff, vl) __riscv_vid_v_u32m8_m(mask, vl)
+#define vid_v_u64m8_m(mask, maskedoff, vl) __riscv_vid_v_u64m8_m(mask, vl)
+#define vadd_vx_u32m8 __riscv_vadd_vx_u32m8
+#define vadd_vx_u64m8 __riscv_vadd_vx_u64m8
+#define vadd_vx_u32m8_m(mask, maskedoff, op1, op2, vl) __riscv_vadd_vx_u32m8_m(mask, op1, op2, vl)
+#define vadd_vx_u64m8_m(mask, maskedoff, op1, op2, vl) __riscv_vadd_vx_u64m8_m(mask, op1, op2, vl)
+#endif
+
 #endif
diff --git a/kernel/riscv64/KERNEL b/kernel/riscv64/KERNEL
index 68d68b5..b94dd3d 100644
--- a/kernel/riscv64/KERNEL
+++ b/kernel/riscv64/KERNEL
@@ -1,3 +1,35 @@
+SDOTKERNEL   = dot_vector.c
+DDOTKERNEL   = dot_vector.c
+
+SAXPYKERNEL  = axpy_vector.c
+DAXPYKERNEL  = axpy_vector.c
+
+SNRM2KERNEL  = nrm2_vector.c
+DNRM2KERNEL  = nrm2_vector.c
+
+SGEMVNKERNEL = gemv_n_vector.c
+DGEMVNKERNEL = gemv_n_vector.c
+SGEMVTKERNEL = gemv_t_vector.c
+DGEMVTKERNEL = gemv_t_vector.c
+
+SASUMKERNEL  = asum_vector.c
+DASUMKERNEL  = asum_vector.c
+
+SSCALKERNEL  = scal_vector.c
+DSCALKERNEL  = scal_vector.c
+
+SCOPYKERNEL  = copy_vector.c
+DCOPYKERNEL  = copy_vector.c
+
+SAXPBYKERNEL = axpby_vector.c
+DAXPBYKERNEL = axpby_vector.c
+
+ISAMAXKERNEL = iamax_vector.c
+IDAMAXKERNEL = iamax_vector.c
+
+ISAMINKERNEL = iamin_vector.c
+IDAMINKERNEL = iamin_vector.c
+
 ifndef SCABS_KERNEL
 SCABS_KERNEL	= ../generic/cabs.c
 endif
@@ -27,4 +59,3 @@ ifndef ZGEMM_BETA
 ZGEMM_BETA = ../generic/zgemm_beta.c
 endif

-
